builder/virtualbox/vm/step_create_snapshot.go

```diff
@@ -0,0 +1,74 @@
+package vm
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	vboxcommon "github.com/hashicorp/packer/builder/virtualbox/common"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+type StepCreateSnapshot struct {
+	Name           string
+	TargetSnapshot string
+}
+
+func (s *StepCreateSnapshot) Run(_ context.Context, state multistep.StateBag) multistep.StepAction {
+	driver := state.Get("driver").(vboxcommon.Driver)
+	ui := state.Get("ui").(packer.Ui)
+	if s.TargetSnapshot != "" {
+		time.Sleep(10 * time.Second) // Wait after the Vm has been shutdown, otherwise creating the snapshot might make the VM unstartable
```

Comment:
Any time we use a sleep to wait for some state change to occur, I get nervous -- is there a better way of testing that the VM is shut down? Is there a reason why a driver.IsRunning() check is not sufficient?

Reply:
Megan, I rewrote the function to create the (final) snapshot. After analysing the code and your hint about driver.IsRunning() I discovered that Step_Shutdown.go firstly only gently waits for the VM to shutdown only if a shutdown command is specified. Otherwise the step simply initiates a Stop on the VM and continue. So the shutdown wasn't realiable in all circumstances. I moved the code for waiting on the VM out of the block for the shudtown command, so that the step realiably shuts down a VM. Secondly the config.go of this builder adds a defaults value of 2 seconds for the post_shutdown_delay which was needed to prevent corruption of the VM while creating a new snapshot if not all locks are released.

Reply:
> Any time we use a sleep to wait for some state change to occur, I get nervous -- is there a better way of testing that the VM is shut down? Is there a reason why a driver.IsRunning() check is not sufficient?

Hi Megan ... I took the chance to see whether other code inside the VirtualBox builder is using Sleep to wait for a status change of a VM and I've found what I'm looking for:

https://github.com/hashicorp/packer/blob/903a4220201462007c1f87016fe840331f7dc1e7/builder/virtualbox/common/driver_4_2.go#L157

https://github.com/hashicorp/packer/blob/903a4220201462007c1f87016fe840331f7dc1e7/builder/virtualbox/common/step_export.go#L51

I think the code inside `driver_4_2.go` can't be changed but perhaps the code for waiting inside `step_export.go` should be rewritten according to your comment.

Maybe `driver_4_2.go` should get an additional method like `EnsureStopped(timeout)` because waiting for a VM to be stopped is a common task and we shouldn't distribute  the required code all over the place and thus lowering maintainablity.